### PR TITLE
Add tester functionality to package

### PR DESCRIPTION
**Pull request by wigging**
Add tester functionality to the torchdet package along with new package dependencies. Also add relevant tester components for the `torchdet test` subcommand.

### CODE REVIEW

**Review by asedova — APPROVED** [review]
moving forward! we still have the interactive stuff happening. do we have a final decision on that? @sanjif-shanmugavelu we are not sure we need the interactive mode but if you think there's a particular reason for it let's discuss. I find it can be a hindrance to batched scaled-up runs of codes.

**Comment by sanjif-shanmugavelu** [comment]
@asedova @wigging do we still want the option for the use to specify which operations to test? I implemented this in an interactive mode, but we could just let the user put all the operations they want to test into a csv and then pass it to the tester? I am not married to either option, it really depends on how we want people to interact with the codebase really

**Comment by asedova** [comment]
> @asedova @wigging do we still want the option for the use to specify which operations to test? I implemented this in an interactive mode, but we could just let the user put all the operations they want to test into a csv and then pass it to the tester? I am not married to either option, it really depends on how we want people to interact with the codebase really

Yes--definitely-- we can have them input a function name or a csv. We definitely want them to be able to chose a particular function, and also-- to be able to run the tester independent of any linter runs, if desired. Which would mean there would be no output list, just the supported functions for testing. Gavin is going to add the available test names to a help menu output, so they can run a help option and get those names to copy-paste into input args for the testing tool.

**Comment by sanjif-shanmugavelu** [comment]
ok nice, thanks Gavin. In the meantime I can remove the interactive option, it's as simple as not calling a function. I already have a flow where we can accept a csv or a list of names in the argparse so that's good

**Comment by asedova** [comment]
>> In the meantime I can remove the interactive option
@sanjif-shanmugavelu , @wigging may already be doing that now!

**Comment by wigging** [comment]
So far I have made the following changes:
- Renamed file_linter.py to linter.py and benchmark_script.py to benchmarks.py
- Removed argparse and input related code in benchmarks.py
- Added `--valid` option to print out valid function names for the test subcommand
- Pinned NumPy version to less than version 2.0, I had trouble running PyTorch with latest version of NumPy
- Used `PYTORCH_DEVICE = "cpu"` in benchmarks.py because I don't have CUDA right now, this should actually be configured via command line argument

**Comment by wigging** [comment]
When I run `torchdet test AvgPool3d` I get several warnings from SciPy stats function about precision loss (see below). Does anyone else get these warnings?

```text
Benchmarking AvgPool3d...
----------Benchmarking <class 'torch.nn.modules.pooling.AvgPool3d'>----------
/Users/miniconda3/envs/torchdet/lib/python3.12/site-packages/scipy/stats/_axis_nan_policy.py:573: RuntimeWarning: Precision loss occurred in moment calculation due to catastrophic cancellation. This occurs when the data are nearly identical. Results may be unreliable.
  res = hypotest_fun_out(*samples, **kwds)
/Users/miniconda3/envs/torchdet/lib/python3.12/site-packages/scipy/stats/_axis_nan_policy.py:573: RuntimeWarning: Precision loss occurred in moment calculation due to catastrophic cancellation. This occurs when the data are nearly identical. Results may be unreliable.
  res = hypotest_fun_out(*samples, **kwds)
/Users/miniconda3/envs/torchdet/lib/python3.12/site-packages/scipy/stats/_axis_nan_policy.py:573: RuntimeWarning: Precision loss occurred in moment calculation due to catastrophic cancellation. This occurs when the data are nearly identical. Results may be unreliable.
  res = hypotest_fun_out(*samples, **kwds)
```

**Comment by sanjif-shanmugavelu** [comment]
I’ve always seen those warning, I believe that’s expected when you’re comparing very similar tensors. I would ignore this, at least for now

**Comment by wigging** [comment]
In kernels/benchmark.py there is `from torchdet.kernels.utils import *`. What is this importing and where is it being used in the file?

**Comment by sanjif-shanmugavelu** [comment]
all the dataclasses and the ```initialise_weights``` function is being imported. The dataclasses are used for the looping through hyperparameters/parameters logic while ```initialise_weights``` sets the weights of ```nn.Modules``` according to a predefined ```torch.init``` initialisation scheme.

On another note, we have used UK English (s instead of z for example), does that matter for the submission?

**Comment by wigging** [comment]
I don't see any data classes or `initialize_weights` function in the kernels/benchmark.py file. I'm looking at this file: https://github.com/minnervva/torchdetscan/blob/det-test/src/torchdet/kernels/benchmark.py. It doesn't look like anything in that file is obtained from the import statement `from torchdet.kernels.utils import *`. Deleting this statement seems to have no effect on the program.

**Comment by sanjif-shanmugavelu** [comment]
I made the following change:

- ```46f0b01``` added logic for checking variability in backward calls

**Comment by sanjif-shanmugavelu** [comment]
sorry Gavin, I was looking at the ```kernels/utils.py``` file. I believe you're right, there should be no need for the ```from torchdet.kernels.utils import *``` import. 

All of the dataclasses and the single function from there are used in the higher level individual operation files, ```kernels/avg_pool.py``` for example. I suspect this ```from torchdet.kernels.utils import *``` import is not even needed in the ```benchmarks.py``` file.

I say this because the relevant kernel.utils artifacts are imported in the individual operation files.

```/torchdetscan/src/torchdet/kernels/avg_pool.py```

```
from torchdet.kernels.utils import (
    HyperParamLoop,
    HyperParams,
    LoopParams,
    Params,
    initialise_weights,
```

and this is propogated downwards

**Comment by sanjif-shanmugavelu** [comment]
Gavin, I am looking through the code again now, with the hopes of adding more operations. I will try to keep an eye on unnecessary imports like the above. I need to familiarise myself a bit more with the codebase

**Comment by sanjif-shanmugavelu** [comment]
gotcha, I can create another branch, or just keep things locally for now

**Comment by wigging** [comment]
I fixed the imports, as in I removed all the `import *` statements. I ran a few examples with `torchdet scan` and `torchdet test` and everything seems fine. But someone else should do some runs too and make sure nothing fails. If everything works, I suggest we go ahead and merge with the main branch.

**Comment by sanjif-shanmugavelu** [comment]
Looks good to me, I think the only clear thing to add is to remove the interactive mode, but that's another story

**Comment by wigging** [comment]
It should be straight-forward to remove the interactive mode. Maybe replace it with an optional argument like
`--select 2,4,5` or just remove it completely.

**Comment by sanjif-shanmugavelu** [comment]
I like the optional argument ```--select 2,4,5``` idea, I can work on that if everyone is happy. Won't be too hard and I think it would save the user from having to manually edit the csv, or create a new one just for the test.

Right now I'm just focusing on adding op support, which I can add once this PR is sorted

**Comment by wigging** [comment]
@sanjif-shanmugavelu I'm going to push up a few more changes then I think we should be good to merge this pull request.

**Comment by wigging** [comment]
I removed the interactive prompt for user input and replaced it with a `--select` argument. A user would run the following `torchdet test output.csv --select 2,4,5` to select certain functions in the CSV file.

This needs to be improved because there's nothing to prevent the user from selecting out-of-bounds for the list of functions; for example, doing `--select 2,4,5,10` doesn't work if the CSV file only lists five functions. Running all the function s in the CSV file also needs to be handled; for example, if `--select` is not given then run all the functions listed in the CSV file. All of this can be fixed in another pull request.

**Comment by sanjif-shanmugavelu** [comment]
nice! I think that could be pretty sraighforward, for example we could add an ```if else``` statement to filter the functions depending on if the ```--select``` flag is present

```
if args.select: # not sure about syntax here
           selected_functions = [functions[int(i) - 1] for i in args.select.split(",") if i.isdigit()]
else:
           selected_functions = functions
```

**Comment by wigging** [comment]
Ok, let me make the change. To be continued...

**Comment by wigging** [comment]
> Running all the functions in the CSV file also needs to be handled; for example, if --select is not given then run all the functions listed in the CSV file.

Done in recent commit.

**Review by markcoletti — APPROVED** [review]
I'm concerned about the paucity of comments and docstrings, but we're in crunch mode, so we can add those later when we have time.  I see many "import *" statements, but I believe Gavin has addressed those elsewhere.

**Comment by wigging** [comment]
@markcoletti The `import *` statements should be gone. Where are you seeing them? Keep in mind that `test.py` will be removed soon so I didn't fix anything in that file. And yes, I agree there are a lack of docstrings and comments. But I don't think that's top priority right now. But since you bring that up, there are also no tests (pytest), no code linter checks (ruff), no formatter checks (ruff), and no GitHub Actions setup to run tests and enforce code style and format for pull requests. These things should have been setup on day one of the project and would be very helpful right now.

**Comment by markcoletti** [comment]
> @markcoletti The `import *` statements should be gone. Where are you seeing them? 

I saw them in the "added lines" and just assumed that you had nuked them in a later version given the comments here and on Slack.

> ...  And yes, I agree there are a lack of docstrings and comments. But I don't think that's top priority right now. 

I agree, hence why I wrote "we're in crunch mode, so we can add those later when we have time."  In any case, I tend to write docstrings and comments while I code so that I have design and implementation thoughts fresh in mind.  (And, sometimes, the very act of writing a docstring or comment can change my mind on how to do things.) I did see some code wondering what the purpose was and why things were the way they were, but at this point we're in flight, and you two seem to have things well in hand.  

> But since you bring that up, there are also no tests (pytest), no code linter checks (ruff), no formatter checks (ruff), and no GitHub Actions setup to run tests and enforce code style and format for pull requests. These things should have been setup on day one of the project and would be very helpful right now.

I would gently disagree with some of what you wrote here given the context of the project and code.  There was informal test driven development (TDD) that yielded example code with the intent of making them more formal unit tests if time allowed; but they nonetheless sufficed to ensure that things worked as intended as they provided ready examples as a pleasant side-effect. I use PyCharm, which has built-in linter and format checks, as do most modern IDEs.  (Though I have some disagreements with what I consider draconian spacing requirements of PEP8, but that's just me.  :) ). This is a simple enough project (at least from the perspective of the linter) that coding guidelines are a bit overkill.  In my experience, those are things negotiated with the development team with the intent of formalizing the code for open-source; in a way, we're in the initial stages of that process now.  E.g., I observe stylistic differences in naming (sometimes using "func" in naming functions, sometimes not), or having a columnar orientation for writing function signatures vs. minimizing the rows used), among other differences.  At some point these styles will have to be homogenized, but I think we can all agree we've got bigger fish to fry given the looming 9/30 deadline.

In any case, the branch is merged into `main` and I'm struggling to find a good place I can jump in without disrupting ongoing development, but that's a discussion best hand on Slack. 😁
